src/pytest_dbfixtures/port.py

```diff
@@ -29,56 +31,59 @@ def __init__(self, ports):
 
     def __str__(self):
         return ('Unknown format of ports: %s.\n'
-                'You should provide an exact port, ports range "4000-5000"'
-                'or a comma-separated ports list "4000,5000,6000-8000".'
+                'You should provide a ports range "[(4000-5000)]"'
+                'or a comma-separated ports set "[{4000,5000,6000,8000}]"'
+                'or both "[(20000, 30000), {48889, 50121}]"'
                 % self.ports)
 
 
 def get_port(ports):
     """
     Retuns a random available port. If there's only one port passed
     (e.g. 5000 or '5000') function does not check if port is available.
+    it there's -1 passed as an argument, function returns None.
     When a range or list of ports is passed `port_for` external package
     is used in order to find a free port.
 
-    :param int|str ports: e.g. 3000, '3000', '3000-3100', '3000,3002', '?'
+    :param int|str|list of tuples and/or sets and/or ints:
+        e.g. 3000, '3000', (3000,3005), {3001, 3005}
+        [(20000, 30000), {48889, 50121}]
+        [(20000, 30000), {48889, 50121}, 3012]
     :returns: a random free port
+    :raises: ValueError
     """
+    if ports == -1:
+        return None
+    elif not ports:
+        return port_for.select_random(None)
+
     try:
         return int(ports)
-    except ValueError:
+    except TypeError:
         pass
 
-    return port_for.select_random(parse_ports(ports))
+    ports_set = set()
 
+    try:
+        if isinstance(ports, (set, frozenset)):
+            ports_set = ports
+        elif isinstance(ports, tuple):
+            ports_set = port_for.utils.ranges_to_set([ports])
+        elif isinstance(ports, list):
+            ranges = port_for.utils.ranges_to_set(filter_by_type(ports, tuple))
+            nums = set(filter_by_type(ports, int))
+            sets = reduce(lambda x, y: x | y, (filter_by_type(
+                ports, (set, frozenset))), set())
 
-def parse_ports(ports):
-    """
-    :param str ports: e.g. '3000', '3000-3100', '3000,3002', '?'
-    :returns: ports set reflecting specifified ports list/range.
-    :rtype set
-    """
+            ports_set = ports_set.union(ranges, sets, nums)
+    except:
```

Comment:
That might hide other exceptions not really related to InvalidPortsDefnition. especially that we do not preserve the original exception cause here.